Adds charger to system output limits

rriveramcrus opened the pull request:
This PR introduces two new properties to configure a notification to be sent to the system in the event of excess battery discharge or a sudden sag in the system output rail.

As a bonus this PR also includes a minor fix to an error introduces in 7d51b7cdc14b56e477e1cbcf2f4f2751f9c6b55b in the documentation for `CHARGER_PROP_INPUT_CURRENT_NOTIFICATION`

rriveramcrus committed:
drivers: charger: Corrects typo in property documentation

Corrects expected type mentioned in the documentation of the
CHARGER_PROP_INPUT_CURRENT_NOTIFICATION.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
drivers: charger: Adds CHARGER_PROP_DISCHARGE_CURRENT_NOTIFICATION

Adds CHARGER_PROP_DISCHARGE_CURRENT_NOTIFICATION property
to notify the system of a excess battery discharge event.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
drivers: charger: Adds CHARGER_PROP_SYSTEM_VOLTAGE_NOTIFICATION_UV

Adds a CHARGER_PROP_SYSTEM_VOLTAGE_NOTIFICATION_UV property to
issue a notification to the system when the system voltage falls.

This system voltage is provided by the charger to the system and can
suddenly dip during peak loads.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

bbilas commented:
Do you think it is possible to extend the charger sample with these features?

rriveramcrus commented:
> Do you think it is possible to extend the charger sample with these features?

It can be added. These properties are not common to all chargers, so I can have the charger sample continue if `ENOTSUP` is returned.

rriveramcrus committed:
samples: charger: Moves sample charger app under samples/drivers/

Moves the charger API sample app from samples/charger/ to
samples/drivers/chargers/ as mentioned in issue #65032.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
samples: charger: Demos new output limit properties

Adds use of CHARGER_PROP_DISCHARGE_CURRENT_NOTIFICATION and
CHARGER_PROP_SYSTEM_VOLTAGE_NOTIFICATION_UV to the charger
sample application.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

kartben commented on samples/drivers/charger/src/main.c:
Thanks for the move! It would really nice to add a README.rst, since right now this sample does not show up in the documentation. 
Using the `:zephyr:code-sample:` directive will also allow to automatically cross reference the sample from the `charger_interface` API reference section of https://docs.zephyrproject.org/latest/hardware/peripherals/charger.html :)

rriveramcrus replied:
will do! ack

rriveramcrus replied:
done a5f4442ab18e2cfe5dd1ffaa8544ca590d0a48f5

kartben commented on samples/drivers/charger/src/main.c:
```diff
@@ -32,6 +32,28 @@ int main(void)
 
 	printk("Found device \"%s\", getting charger data\n", chgdev->name);
 
+	val.discharge_current_notification.current_ua = 5000000;
```
Should such values be Kconfigs, and are these default values ok to have? To put it differently: are these hard coded value realistically (and safely?) applicable to any setup and type of battery the person testing will be using?

rriveramcrus replied:
yeah that sounds like a better as more high power chargers get merged in. I'll add some Kconfig options.

rriveramcrus replied:
I think I'm close here (a1b86115c544d2632eb1baa8b589ce45a2bb9de9), but I am running into a few issue with the config symbols. I'll try and resolve it on the next push.

kartben reviewed (commented):
I won't make this a "request for change" but it would be really nice to finally have a README for this sample. Not having it show up in the documentation makes me a sad puppy :)

rriveramcrus committed:
samples: charger: Documents charger sample app

Creates basic documentation for the chagrer sample application.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

fabiobaltieri commented:
@rriveramcrus can you add the sample path in the Charger area in the maintainers file as well?

rriveramcrus committed:
MAINTAINERS: Adds sample path to the charger area

Adds the path to the sample application in the charger API
area of the MAINTAINERS file.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus commented:
> @rriveramcrus can you add the sample path in the Charger area in the maintainers file as well?

Done, in a0852cc1f4fdc1f6af032b3ec31f73db72c9a21f

kartben commented on drivers/charger/Kconfig:
```diff
@@ -19,6 +19,37 @@ config CHARGER_INIT_PRIORITY
 	help
 	  Battery charger initialization priority.
 
+config CHARGER_HAS_DISCHAGRE_CURRENT_NOTIFICATIONS
```
s/DISCHAGRE/DISCHARGE/g

rriveramcrus replied:
fixed

aaronemassey commented:
> This PR introduces two new properties to configure a notification to be sent to the system in the event of excess battery discharge or a sudden sag in the system output rail.
> 
> As a bonus this PR also includes a minor fix to an error introduces in 7d51b7c in the documentation for CHARGER_PROP_INPUT_CURRENT_NOTIFICATION

It's best to make PRs as small as possible. It makes things easier to review,
especially with GitHub UI.

That being said: LGTM but I think we talked about removing me as a maintainer
for this subsystem (don't have the cycles) so reassigning to @GRobertZieba as a
Charging subsystem collaborator. I'll push out a PR similar to the closed
<https://github.com/zephyrproject-rtos/zephyr/pull/67404> to remove myself as a
maintainer and add @GRobertZieba as a collaborator.

kartben reviewed (approved):
sample README could use more love but this can be addressed in a follow-up PR

fabiobaltieri commented on samples/drivers/charger/src/main.c:
```diff
@@ -32,6 +32,29 @@ int main(void)
 	printk("Found device \"%s\", getting charger data\n", chgdev->name);
 
+#ifdef CONFIG_CHARGER_DISCHARGE_CURRENT_NOTIFICATIONS
+	val.discharge_current_notification.current_ua =
+		CONFIG_APP_DISCHARGE_CURRENT_NOTIFICATION_THRESHOLD_UA;
+	val.discharge_current_notification.severity = CHARGER_SEVERITY_WARNING;
+	val.discharge_current_notification.duration_us =
+		CONFIG_APP_DISCHARGE_CURRENT_NOTIFICATION_DURATION_US;
```
Hey does this really have to be part of the runtime API? At first looks it seems like a device specific configuration that should just be setup for the specific device at init time, the parameters be in the binding and never touched again.

rriveramcrus replied:
It is necessary for ECs managing charging/discharging on x86 platforms. If this were an absolute maximum trip point that protected the battery pack, I would agree that yes it could be set and forgotten at init. 

However the purpose of these notification events is not to protect the battery but rather to prevent a system brownout by having the charger notify the system of excess power being drawn from the battery. The threshold at which the charger should notify the system can vary depending on the known SOH/SOC of the battery.

Examples of conditions that could reduce maximum current you could draw from the battery aside from SOC:
- Heat
- Cycle count
- Cold
- The absence of an external power source
- The time between turbo event (load frequency)

This feature is supported on the TI, Intersil, and Southchip chargers targeting x86 platforms. I could see a case for using a Kconfig to enable/disable the `CHARGER_PROP_*_NOTIFICATION` props.
